### PR TITLE
tests: added "fleetctl --tunnel" tests

### LINKED ISSUE #1484: tests: create new tests
for these issues: https://github.com/coreos/fleet/issues/565 https://github.com/coreos/fleet/issues/1425 https://github.com/coreos/fleet/issues/1455 (`Error running remote command: timed out while initiating SSH connection`)
and probably for this one https://github.com/coreos/fleet/issues/1402

### PR DESCRIPTION
**Pull request by kayrus**
adding tests which will cover #565 #1425 #1455
closes #1484

### CODE REVIEW

**Comment by kayrus** [comment]
Looks like `WaitForNActiveUnits` doesn't wait for N active units, which are templates. have to fix that.

**Comment by kayrus** [comment]
Yet another test for #565:

```sh
$ export REMOTE_HOST=localhost
$ for i in {1..10}; do (ssh-keygen -t rsa -b 4096 -f $HOME/.ssh/id_rsa_$i -N ''; ssh-add $HOME/.ssh/id_rsa_$i)& done; wait; echo "done"
$ fleetctl -strict-host-key-checking=false --tunnel $REMOTE_HOST list-machines
Unable to initialize client: failed initializing SSH client: ssh: handshake failed: ssh: unexpected message type 1 (expected 52)
```

**Comment by kayrus** [comment]
Trying to find other way to fix this bug. https://github.com/coreos/fleet/pull/1492/commits/ea415adbb229624f9f632f7c69a4829036a1a17d is only workaround which doesn't explain cause of the problem

**Comment by kayrus** [comment]
@jonboulle if you accept this fix, let's merge this PR. and I'll try to find proper solution later within new issue.

File: functional/platform/nspawn.go [hunk 1]
```diff
@@ -228,14 +252,14 @@ func (nc *nspawnCluster) prepCluster() (err error) {
...
-	cmd := fmt.Sprintf("mkdir -p %s/opt/fleet", dir)
+func (nc *nspawnCluster) insertBin(src string, dst string) error {
+	cmd := fmt.Sprintf("mkdir -p %s/opt/fleet", dst)
 	if _, _, err := run(cmd); err != nil {
 		return err
 	}
 
-	fleetdBinDst := path.Join(dir, "opt", "fleet", "fleetd")
-	return copyFile(fleetdBinPath, fleetdBinDst, 0755)
+	BinDst := path.Join(dst, "opt", "fleet", path.Base(src))
```
Comment (jonboulle) [inline functional/platform/nspawn.go]: nit, binDst

File: functional/tunnel_test.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	tmp.Close()
+	defer syscall.Unlink(tmp.Name())
+
+	khFile := tmp.Name()
+
+	// Launch one unit
+	if stdout, stderr, err := cluster.FleetctlWithInput(m0, "yes", fmt.Sprintf("--tunnel=%s", m0.IP()), "--strict-host-key-checking=true", fmt.Sprintf("--known-hosts-file=%s", khFile), "start", "fixtures/units/hello.service"); err != nil {
```
Comment (jonboulle) [inline functional/tunnel_test.go]: This line is waaaay too long to be readable, can you split out the arguments on separate lines?

**Comment by jonboulle** [comment]
Couple of nits but LGTM overall!

**Comment by kayrus** [comment]
@tixxdz any remarks or objections?

File: ssh/ssh.go [hunk 3]
```diff
@@ -30,10 +30,14 @@ import (
 type SSHForwardingClient struct {
 	agentForwarding bool
 	*gossh.Client
+	agentForwardingEnabled bool
```
Comment (tixxdz) [inline ssh/ssh.go]: Ok, one minor: could you please rename this to: authAgentReqSent or agentForwardingSent , but not Enabled, and also put some code comment: "Used to check if we have already sent the request to the ssh channel".

**Comment by tixxdz** [comment]
@kayrus thank you, please if you could change the name of that variable! then will merge.

**Comment by tixxdz** [comment]
@kayrus ok thank you merged, and thanks also for triggering that  "remote: Daily bandwidth rate limit exceeded..." !

**Pull request merged**